x-pack/plugins/alerting/server/lib/trim_recovered_alerts.ts

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger } from '@kbn/logging';
+import { map } from 'lodash';
+import { Alert } from '../alert';
+import { AlertInstanceState, AlertInstanceContext } from '../types';
+
+interface TrimmedRecoveredAlertsResult<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  RecoveryActionGroupIds extends string
+> {
+  trimmedAlertsRecovered: Record<string, Alert<State, Context, RecoveryActionGroupIds>>;
+  trimmedAlertsRecoveredCurrent: Record<string, Alert<State, Context, RecoveryActionGroupIds>>;
+}
+
+export interface TrimRecoveredOpts {
+  key: string;
+  flappingHistory: boolean[];
+}
+
+export function trimRecoveredAlerts<
+  State extends AlertInstanceState,
+  Context extends AlertInstanceContext,
+  RecoveryActionGroupIds extends string
+>(
+  logger: Logger,
+  recoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupIds>> = {},
+  currentRecoveredAlerts: Record<string, Alert<State, Context, RecoveryActionGroupIds>> = {},
+  maxAlerts: number
+): TrimmedRecoveredAlertsResult<State, Context, RecoveryActionGroupIds> {
+  const alerts = map(recoveredAlerts, (value, key) => {
+    return {
+      key,
+      flappingHistory: value.getFlappingHistory() || [],
+    };
+  });
+  const earlyRecoveredAlerts = getEarlyRecoveredAlerts(logger, alerts, maxAlerts);
+  // Dropping the "early recovered" alerts for now.
+  // In #143445 we will want to recover these alerts and set flapping to false
+  earlyRecoveredAlerts.forEach((alert) => {
+    delete recoveredAlerts[alert.key];
+    delete currentRecoveredAlerts[alert.key];
+  });
+  return {
+    trimmedAlertsRecovered: recoveredAlerts,
+    trimmedAlertsRecoveredCurrent: currentRecoveredAlerts,
```

Comment:
This still doesn't make sense to me, maybe I'm missing something.  What I wanted to make sure happens is that the object containing the flapping info for alerts that have recovered recently but aren't active right now, doesn't go over the max alerts circuit breaker.  

I'm still confused by the names - I believe the "current" version is the alerts that JUST recovered, and the non-"current" version is the alerts we are tracking for flapping.

I don't think we want to remove the `currentRecoveredAlerts` items at all.  We just want to stop tracking the flapping history of alerts that go over the circuit breaker.  Actually, I don't think it even needs to get passed in, just the `recoveredAlerts`.

I guess we'll have some interesting cases.  I think any alert in `currentRecoveredAlerts` is also in `recoveredAlerts`, butI wouldn't expect an alert that JUST recovered to actually "lose" the circuit breaker, and we'd stop tracking it's history.  But maybe that can happen for an edge case.  If it does, we still want to keep alert in `currentRecoveredAlerts` so it'll run it's actions. We'd just lose it's flapping history till it went active again.